libs/ui/src/hooks/smartcard_auth/use_dipped_smartcard_auth.ts

```diff
@@ -51,130 +66,157 @@ type SmartcardAuthAction =
   | { type: 'log_out' }
   | { type: 'bootstrap_admin_session'; electionHash: string };
 
-function validateCardUser(user?: User): Result<void, LoggedOut['reason']> {
-  if (!user) return err('invalid_user_on_card');
-  if (!(user.role === 'admin' || user.role === 'superadmin')) {
+function validateCardUser(
+  user: Optional<User>,
+  scope: DippedSmartcardAuthScope
+): Result<void, LoggedOut['reason']> {
+  if (!user) {
+    return err('invalid_user_on_card');
+  }
+
+  if (!['superadmin', 'admin'].includes(user.role)) {
     return err('user_role_not_allowed');
   }
+
+  if (user.role === 'admin') {
+    if (!areVvsg2AuthFlowsEnabled()) {
+      return ok();
+    }
+    if (!scope.electionDefinition) {
+      return scope.allowAdminsToAccessUnconfiguredMachines
+        ? ok()
+        : err('machine_not_configured');
+    }
+    if (user.electionHash !== scope.electionDefinition.electionHash) {
+      return err('admin_wrong_election');
+    }
+  }
+
   return ok();
 }
 
-function smartcardAuthReducer(
-  previousState: SmartcardAuthState,
-  action: SmartcardAuthAction
-): SmartcardAuthState {
-  switch (action.type) {
-    case 'card_read': {
-      const newAuth = ((): AuthState => {
-        switch (previousState.auth.status) {
-          case 'logged_out': {
-            switch (action.cardSummary.status) {
-              case 'no_card':
-                return { status: 'logged_out', reason: 'machine_locked' };
-
-              case 'error':
-                return { status: 'logged_out', reason: 'card_error' };
-
-              case 'ready': {
-                const user = parseUserFromCardSummary(action.cardSummary);
-                const validationResult = validateCardUser(user);
-                if (validationResult.isOk()) {
-                  assert(
-                    user &&
-                      (user.role === 'superadmin' || user.role === 'admin')
-                  );
-                  return { status: 'checking_passcode', user };
+function smartcardAuthReducer(scope: DippedSmartcardAuthScope) {
```

Comment:
I recommend reviewing this section with [`?w=1`](https://github.com/votingworks/vxsuite/pull/2224/commits/ef143331bf6b4e2c247a44965e1309056bce49fa?w=1) since the bulk of the diff is a whitespace change